src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -85,15 +93,102 @@ public IReadOnlyList<string> AllPresentEndogenousStationaryMoleculeNames() =>
 
       private void performMerge()
       {
-         cacheBuilders(x => x.Reactions, _reactions);
-         cacheBuilders(x => x.Molecules, _molecules);
-         cacheMoleculeDependentBuilders(x => x.PassiveTransports, _passiveTransports);
-         cacheMoleculeDependentBuilders(x => x.Observers, _observers);
-         cacheParameterValueBuilders(x => x.SelectedParameterValues, _parameterValues);
-         cacheInitialConditions();
+         mergeBuilders(x => x.Reactions, _reactions, mergeReactions);
+         mergeBuilders(x => x.Molecules, _molecules, mergeMolecules);
+         mergeBuilders(x => x.PassiveTransports, _passiveTransports, mergeTransports);
+         mergeBuilders(x => x.Observers, _observers, mergeObservers);
+         mergeParameterValueBuilders(x => x.SelectedParameterValues, _parameterValues);
+         mergeInitialConditions();
          cacheEntities();
       }
 
+      private void mergeBuilders<T>(Func<Module, IBuildingBlock<T>> propAccess, CacheByName<T> cache, Action<T, BuilderSource<T>> mergeStrategyAction) where T : class, IBuilder, IEntity
+      {
+         var analyzedMerges = analyzeBuilderMerges(propAccess);
+
+         foreach (var mergeInfo in analyzedMerges)
+         {
+            var (finalBuilder, buildingBlock) = mergeInfo.BaseBuilder;
+
+            if (mergeInfo.RequiresClone)
+            {
+               finalBuilder = cloneBuilder(finalBuilder);
+               foreach (var (sourceBuilder, sourceBuildingBlock) in mergeInfo.BuildersToMerge)
+               {
+                  //Clone the source builder if needed to prevent cross-contamination during sequential merges
+                  //When merging multiple builders, earlier merged builders could be affected by later merges through shared references
+                  var sourceBuilderToMerge = mergeInfo.RequiresMergeClone ? cloneBuilder(sourceBuilder) : sourceBuilder;
+                  var mergedBuilderSource = new BuilderSource<T>(sourceBuilderToMerge, sourceBuildingBlock);
+                  tryMergeContainers(finalBuilder, mergedBuilderSource);
+                  mergeStrategyAction(finalBuilder, mergedBuilderSource);
+               }
+            }
+
+            cache.Add(finalBuilder);
+            AddToBuilderSource(finalBuilder, buildingBlock);
+         }
+      }
+
+      private T cloneBuilder<T>(T builder) where T : class, IBuilder
+      {
+         //we will need to make sure we keep the Id the same somehow to ensure proper tracking of entities. TBD
```

Comment:
I think that we will need to ensure that we KEEP The ids of the cloned entities. Because we use the id to find the element in the configuration but if we clone, it's gone! (this is to ensure we can Jump to Source only)

Reply:
Meaning this should only be used from the context of a warning or error message in the notifications pane?